server/repository/src/db_diesel/indicator_value_row.rs

```diff
@@ -0,0 +1,122 @@
+use super::{
+    ChangeLogInsertRow, ChangelogRepository, ChangelogTableName, RowActionType, StorageConnection,
+};
+use crate::{repository_error::RepositoryError, Delete, Upsert};
+use diesel::prelude::*;
+
+table! {
+    indicator_value (id) {
+        id -> Text,
+        customer_name_link_id -> Text,
+        supplier_store_id -> Text,
+        period_id -> Text,
+        indicator_line_id -> Text,
+        indicator_column_id -> Text,
+        value -> Text,
+    }
+}
+
+#[derive(Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Default)]
+#[diesel(table_name = indicator_value)]
+pub struct IndicatorValueRow {
+    pub id: String,
+    pub customer_name_link_id: String,
+    pub supplier_store_id: String,
+    pub period_id: String,
+    pub indicator_line_id: String,
+    pub indicator_column_id: String,
+    pub value: String,
+}
+
+pub struct IndicatorValueRowRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> IndicatorValueRowRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        IndicatorValueRowRepository { connection }
+    }
+
+    pub fn upsert_one(&self, row: &IndicatorValueRow) -> Result<i64, RepositoryError> {
+        diesel::insert_into(indicator_value::table)
+            .values(row)
+            .on_conflict(indicator_value::id)
+            .do_update()
+            .set(row)
+            .execute(self.connection.lock().connection())?;
+
+        self.insert_changelog(row, RowActionType::Upsert)
+    }
+
+    pub fn delete(&self, id: &str) -> Result<Option<i64>, RepositoryError> {
+        let old_row = self.find_one_by_id(id)?;
+        let change_log_id = match old_row {
+            Some(old_row) => self.insert_changelog(&old_row, RowActionType::Delete)?,
+            None => {
+                return Ok(None);
+            }
+        };
+
+        diesel::delete(indicator_value::table.filter(indicator_value::id.eq(id)))
+            .execute(self.connection.lock().connection())?;
+        Ok(Some(change_log_id))
+    }
+
+    fn insert_changelog(
+        &self,
+        row: &IndicatorValueRow,
+        action: RowActionType,
+    ) -> Result<i64, RepositoryError> {
+        let row = ChangeLogInsertRow {
+            table_name: ChangelogTableName::IndicatorValue,
+            record_id: row.id.clone(),
+            row_action: action,
+            store_id: Some(row.supplier_store_id.clone()),
+            name_link_id: Some(row.customer_name_link_id.clone()),
```

Comment:
I'm not really sure if I should be populating this or store_id... indicator sync still confusing me haha

Reply:
Both ~

Reply:
It's just confusing because I think the name_link_id is primarily for finding invoices where "other_party_name_id" indicates it's for an active store on the site the site, and "store_id" may belong to a store active on a different site. Though I guess that's similar here... it's just that it's unual for a site to make `change_log` records with an inactive store_id.

Basically I am not 100% here because typically if you add a record record to the change log, the store_id indicates that the record is actually owned by a store on your site. Normally you wouldn't touch a record with a store_id indicating the record is from another site...

Reply:
hmm yeah store_id is for ownership purposes really, i would leave both store_id and name_link_id empty for now. Otherwise might have issues with ChangelogFilter. When we get to do omSupply central, we have more fine grained filtering and can filter by record fields